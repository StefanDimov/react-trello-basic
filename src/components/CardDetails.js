import React from 'react'
import R from 'ramda'
import cardUtils from '../utils/cardUtils'
import { FormGroup, FormControl, ControlLabel } from 'react-bootstrap'

export default class CardDetails extends React.Component {
    constructor(props) {
        super(props)

        this.state = R.clone(props.card)

        this.handleTitleChange = this.handleTitleChange.bind(this)
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleTitleChange(event) {
        this.setState({title: event.target.value})
    }

    handleDescriptionChange(event) {
        this.setState({description: event.target.value})
    }

    handleSubmit(event) {
        event.preventDefault();
        // doesn't do anything
        // just to disable page reload on hitting enter
    }

    componentWillUnmount() {
        // if the card has been changed
        // also when new card: it won't go in if no field has been changed
        if (!R.equals(this.state, this.props.card)) {
            // save new card
            this.props.onCardSave(R.clone(this.state));
        }
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <FormGroup>
                    <ControlLabel>Title</ControlLabel>
                    <FormControl
                        id="card-details-title-input"
                        type="text"
                        value={this.state.title}
                        placeholder="Title..."
                        onChange={this.handleTitleChange} />
                </FormGroup>

                <FormGroup>
                    <ControlLabel>Description</ControlLabel>
                    <FormControl
                        id="card-details-description-input"
                        componentClass="textarea"
                        value={this.state.description}
                        placeholder="Description..."
                        onChange={this.handleDescriptionChange} />
                </FormGroup>
            </form>
        )
    }
}

CardDetails.propTypes = {
    card: cardUtils.cardPropType.isRequired,
    onCardSave: React.PropTypes.func.isRequired
}
