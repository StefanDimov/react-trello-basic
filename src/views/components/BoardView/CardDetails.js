import React from 'react'
import R from 'ramda'
import * as cardUtils from '../../../utils/cardUtils'
import { Row, Col, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap'

/**
 * Component that visualizes a Card's data in inputs. Card's data can be edited.
 * There is a side menu with actions that can be performed on the card as well.
 * @class comp/CardDetails
 * @param {object} card The card data to be visualized
 * @param {function} onCardSave callback which is called with the data when a card needs to be saved
 * @param {function} onCardDelete callback which is called with the data when a card needs to be deleted
 * @param {function} onCardCopy callback which is called with the data when a card needs to be copied
 */
export default class CardDetails extends React.Component {
    constructor(props) { // eslint-disable-line
        super(props)

        this.state = R.clone(props.card)

        this.handleTitleChange = this.handleTitleChange.bind(this)
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    /**
     * Handler for the change event of the title input
     * @private
     * @param {object} event event data passed to the handler
     */
    handleTitleChange(event) {
        this.setState({title: event.target.value})
    }

    /**
     * Handler for the change event of the description input
     * @private
     * @param {object} event event data passed to the handler
     */
    handleDescriptionChange(event) {
        this.setState({description: event.target.value})
    }

    /**
     * Handler for the submit event of the form
     * @private
     * @param {object} event event data passed to the handler
     */
    handleSubmit(event) {
        event.preventDefault()
        // doesn't do anything
        // just to disable page reload on hitting enter
    }

    /**
     * When the component will unmount it will check if the card's data
     * has been changed in any way, if so it will call the onCardSave callback.
     * Also when new card is used (no fields assigned) it won't go in if no
     * field has been changed.
     * @private
     */
    componentWillUnmount() {
        // if the card has been changed
        if (!R.equals(this.state, this.props.card)) {
            // save new card
            this.props.onCardSave(R.clone(this.state))
        }
    }

    render() { // eslint-disable-line
        return (
            <Row>
                <Col sm={9}>
                    <form onSubmit={this.handleSubmit}>
                        <FormGroup>
                            <ControlLabel htmlFor="card-details-title-input">Title</ControlLabel>
                            <FormControl
                                id="card-details-title-input"
                                type="text"
                                value={this.state.title}
                                placeholder="Title..."
                                onChange={this.handleTitleChange} />
                        </FormGroup>

                        <FormGroup>
                            <ControlLabel htmlFor="card-details-description-input">Description</ControlLabel>
                            <FormControl
                                id="card-details-description-input"
                                componentClass="textarea"
                                value={this.state.description}
                                placeholder="Description..."
                                onChange={this.handleDescriptionChange} />
                        </FormGroup>
                    </form>
                </Col>

                <Col sm={3}>
                    <ControlLabel>Actions</ControlLabel>
                    <Button
                        id="card-copy-button"
                        block={true}
                        onClick={() => this.props.onCardCopy(R.clone(this.state))}
                        bsStyle="primary" bsSize="sm">Copy</Button>

                    <Button
                        id="card-delete-button"
                        block={true}
                        onClick={() => this.props.onCardDelete(R.clone(this.state))}
                        bsStyle="danger" bsSize="sm">Delete</Button>
                </Col>
            </Row>
        )
    }
}

CardDetails.propTypes = {
    card: cardUtils.cardPropType.isRequired,
    onCardSave: React.PropTypes.func.isRequired,
    onCardDelete: React.PropTypes.func.isRequired,
    onCardCopy: React.PropTypes.func.isRequired
}
