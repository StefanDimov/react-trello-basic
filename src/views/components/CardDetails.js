import React from 'react'
import R from 'ramda'
import cardUtils from '../../utils/cardUtils'
import { Grid, Row, Col, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap'

export default class CardDetails extends React.Component {
    constructor(props) {
        super(props)

        this.state = R.clone(props.card)

        this.handleTitleChange = this.handleTitleChange.bind(this)
        this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleTitleChange(event) {
        this.setState({title: event.target.value})
    }

    handleDescriptionChange(event) {
        this.setState({description: event.target.value})
    }

    handleSubmit(event) {
        event.preventDefault();
        // doesn't do anything
        // just to disable page reload on hitting enter
    }

    componentWillUnmount() {
        // if the card has been changed
        // also when new card: it won't go in if no field has been changed
        if (!R.equals(this.state, this.props.card)) {
            // save new card
            this.props.onCardSave(R.clone(this.state));
        }
    }

    render() {
        return (
            <Row>
                <Col sm={9}>
                    <form onSubmit={this.handleSubmit}>
                        <FormGroup>
                            <ControlLabel htmlFor="card-details-title-input">Title</ControlLabel>
                            <FormControl
                                id="card-details-title-input"
                                type="text"
                                value={this.state.title}
                                placeholder="Title..."
                                onChange={this.handleTitleChange} />
                        </FormGroup>

                        <FormGroup>
                            <ControlLabel htmlFor="card-details-description-input">Description</ControlLabel>
                            <FormControl
                                id="card-details-description-input"
                                componentClass="textarea"
                                value={this.state.description}
                                placeholder="Description..."
                                onChange={this.handleDescriptionChange} />
                        </FormGroup>
                    </form>
                </Col>

                <Col sm={3}>
                    <ControlLabel>Actions</ControlLabel>
                    <Button
                        id="card-copy-button"
                        block={true}
                        onClick={() => this.props.onCardCopy(R.clone(this.state))}
                        bsStyle="primary" bsSize="sm">Copy</Button>

                    <Button
                        id="card-delete-button"
                        block={true}
                        onClick={() => this.props.onCardDelete(R.clone(this.state))}
                        bsStyle="danger" bsSize="sm">Delete</Button>
                </Col>
            </Row>
        )
    }
}

CardDetails.propTypes = {
    card: cardUtils.cardPropType.isRequired,
    onCardSave: React.PropTypes.func.isRequired,
    onCardDelete: React.PropTypes.func.isRequired,
    onCardCopy: React.PropTypes.func.isRequired
}
