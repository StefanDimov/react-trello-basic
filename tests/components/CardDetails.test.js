import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import CardDetails from '../../src/components/CardDetails'

describe('CardDetails', () => {

    const titleInputSelector = { id: 'card-details-title-input' }
    const titleDescriptionSelector = { id: 'card-details-description-input' }
    const event = { target: { value: 'some value' } }
    let card
    let onCardSave = jest.fn()

    beforeEach(() => {
        card = { id: 'uniqueId' }
        onCardSave.mockClear()
    })

    describe('rendering', () => {
        it('should render correctly when given empty card', () => {
            const tree = renderer.create(
                <CardDetails card={card} onCardSave={onCardSave} />
            ).toJSON();
            expect(tree).toMatchSnapshot();
        })

        it('should render correctly when given card with title', () => {
            card.title = 'title'
            const tree = renderer.create(
                <CardDetails card={card} onCardSave={onCardSave} />
            ).toJSON();
            expect(tree).toMatchSnapshot();
        })

        it('should render correctly when given card with description', () => {
            card.description = 'description'
            const tree = renderer.create(
                <CardDetails card={card} onCardSave={onCardSave} />
            ).toJSON();
            expect(tree).toMatchSnapshot();
        })

        it('should render correctly when given card with title and description', () => {
            card.title = 'title'
            card.description = 'description'
            const tree = renderer.create(
                <CardDetails card={card} onCardSave={onCardSave} />
            ).toJSON();
            expect(tree).toMatchSnapshot();
        })
    })

    describe('container', () => {
        it('should be a form element', () => {
            const wrapper = shallow(<CardDetails card={card} onCardSave={onCardSave} />)
            expect(wrapper.is('form')).toBe(true)
        })

        it('should only preventDefault on submit', () => {
            const wrapper = shallow(<CardDetails card={card} onCardSave={onCardSave} />)
            const event = { preventDefault: jest.fn() }
            wrapper.simulate('submit', event)
            expect(event.preventDefault.mock.calls.length).toBe(1)
            expect(onCardSave.mock.calls.length).toBe(0)
        })
    })

    describe('title input', () => {
        it('should update state on change', () => {
            const wrapper = shallow(<CardDetails card={card} onCardSave={onCardSave} />)
            expect(wrapper.find(titleInputSelector).prop('value')).toBeUndefined()

            wrapper.find(titleInputSelector).simulate('change', event)
            expect(wrapper.find(titleInputSelector).prop('value')).toBe(event.target.value)
        })
    })

    describe('description input', () => {
        it('should update state on change', () => {
            const wrapper = shallow(<CardDetails card={card} onCardSave={onCardSave} />)
            expect(wrapper.find(titleDescriptionSelector).prop('value')).toBeUndefined()

            wrapper.find(titleDescriptionSelector).simulate('change', event)
            expect(wrapper.find(titleDescriptionSelector).prop('value')).toBe(event.target.value)
        })
    })

    describe('card save logic', () => {
        it('should be called when components unmounts and card has changed', () => {
            const wrapper = shallow(<CardDetails card={card} onCardSave={onCardSave} />)

            // must have a change so onCardSave will be called
            wrapper.find(titleInputSelector).simulate('change', event)
            // should'n te called before unmount
            expect(onCardSave.mock.calls.length).toBe(0)

            wrapper.unmount();
            expect(onCardSave.mock.calls.length).toBe(1)
        })

        it('should not be called if the card has not changed', () => {
            const wrapper = shallow(<CardDetails card={card} onCardSave={onCardSave} />)
            wrapper.unmount();
            expect(onCardSave.mock.calls.length).toBe(0)
        })
    })
})
