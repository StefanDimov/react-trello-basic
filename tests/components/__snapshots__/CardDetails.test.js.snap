exports[`CardDetails rendering should render correctly when given card with description 1`] = `
<form
  onSubmit={[Function]}>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Title
    </label>
    <input
      className="form-control"
      id="card-details-title-input"
      onChange={[Function]}
      placeholder="Title..."
      type="text"
      value={undefined} />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Description
    </label>
    <textarea
      className="form-control"
      id="card-details-description-input"
      onChange={[Function]}
      placeholder="Description..."
      type={undefined}
      value="description" />
  </div>
</form>
`;

exports[`CardDetails rendering should render correctly when given card with title 1`] = `
<form
  onSubmit={[Function]}>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Title
    </label>
    <input
      className="form-control"
      id="card-details-title-input"
      onChange={[Function]}
      placeholder="Title..."
      type="text"
      value="title" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Description
    </label>
    <textarea
      className="form-control"
      id="card-details-description-input"
      onChange={[Function]}
      placeholder="Description..."
      type={undefined}
      value={undefined} />
  </div>
</form>
`;

exports[`CardDetails rendering should render correctly when given card with title and description 1`] = `
<form
  onSubmit={[Function]}>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Title
    </label>
    <input
      className="form-control"
      id="card-details-title-input"
      onChange={[Function]}
      placeholder="Title..."
      type="text"
      value="title" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Description
    </label>
    <textarea
      className="form-control"
      id="card-details-description-input"
      onChange={[Function]}
      placeholder="Description..."
      type={undefined}
      value="description" />
  </div>
</form>
`;

exports[`CardDetails rendering should render correctly when given empty card 1`] = `
<form
  onSubmit={[Function]}>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Title
    </label>
    <input
      className="form-control"
      id="card-details-title-input"
      onChange={[Function]}
      placeholder="Title..."
      type="text"
      value={undefined} />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Description
    </label>
    <textarea
      className="form-control"
      id="card-details-description-input"
      onChange={[Function]}
      placeholder="Description..."
      type={undefined}
      value={undefined} />
  </div>
</form>
`;
